compiler/damlc/tests/src/DA/Test/DamlcUpgrades.hs

```diff
@@ -0,0 +1,525 @@
+              [ ( "daml/MyLib.daml"
+                , unlines
+                      [ "module MyLib where"
+                      , "template A with"
+                      , "    p : Party"
+                      , "    q : Party"
+                      , "  where"
+                      , "    signatory p"
+                      , "    choice C : ()"
+                      , "      controller p"
+                      , "      do pure ()"
+                      ]
+                )
+              ]
+              [ ("daml/MyLib.daml"
+                , unlines
+                      [ "module MyLib where"
+                      , "template A with"
+                      , "    p : Party"
+                      , "    q : Party"
+                      , "  where"
+                      , "    signatory p"
+                      , "    choice C : ()"
+                      , "      controller p, q"
+                      , "      do pure ()"
+                      ]
+                )
+              ]
+        , test
+              "Fails when observers of template choice are changed"
+              (Just "Message: \n\ESC\\[0;91merror type checking template MyLib.A choice C:\n  The upgraded choice C cannot change the definition of observers.")
+              [ ( "daml/MyLib.daml"
+                , unlines
+                      [ "module MyLib where"
+                      , "template A with"
+                      , "    p : Party"
+                      , "    q : Party"
+                      , "  where"
+                      , "    signatory p"
+                      , "    choice C : ()"
+                      , "      observer p"
+                      , "      controller p"
+                      , "      do pure ()"
+                      ]
+                )
+              ]
+              [ ("daml/MyLib.daml"
+                , unlines
+                      [ "module MyLib where"
+                      , "template A with"
+                      , "    p : Party"
+                      , "    q : Party"
+                      , "  where"
+                      , "    signatory p"
+                      , "    choice C : ()"
+                      , "      observer p, q"
+                      , "      controller p"
+                      , "      do pure ()"
+                      ]
+                )
+              ]
+        , test
+              "Fails when template choice changes its return type"
+              (Just "Message: \n\ESC\\[0;91merror type checking template MyLib.A choice C:\n  The upgraded choice C cannot change its return type.")
+              [ ( "daml/MyLib.daml"
+                , unlines
+                      [ "module MyLib where"
+                      , "template A with"
+                      , "    p : Party"
+                      , "  where"
+                      , "    signatory p"
+                      , "    choice C : ()"
```

Comment:
Could you add a test case for C : T, where T is a template, and try to upgrade T as part of the test? This should succeed.

Reply:
Similarly, a test case where a choice input argument is of type Template, and this is upgraded. 

Later TODO: verify whether the engine code successfully handles adding optional arguments to the choice. Moritz said this is already possible, but it would be good to verify whether this is fully correct, and if so, handle it in the compiler accordingly.

Reply:
Fixed

Reply:
https://github.com/digital-asset/daml/issues/17612 for later todo